tidy3d/components/boundary.py

```diff
@@ -0,0 +1,626 @@
+    )
+
+
+# pml types allowed in simulation init
+PMLTypes = Union[PML, StablePML, Absorber, None]
+
+
+# """ boundary specification classes """
+
+# types of boundaries that can be used in Simulation
+
+Symmetry = Literal[0, -1, 1]
+BoundaryEdgeType = Union[
+    Periodic, PECBoundary, PMCBoundary, PML, StablePML, Absorber, BlochBoundary
+]
+
+
+class Boundary(Tidy3dBaseModel):
+    """Boundary conditions at the minus and plus extents along a dimension
+
+    Example
+    -------
+    >>> boundary = Boundary(plus = PML(), minus = PECBoundary())
+    """
+
+    plus: BoundaryEdgeType = pd.Field(
+        Periodic(),
+        title="Plus BC",
+        description="Boundary condition on the plus side along a dimension.",
+    )
+
+    minus: BoundaryEdgeType = pd.Field(
+        Periodic(),
+        title="Minus BC",
+        description="Boundary condition on the minus side along a dimension.",
+    )
+
+    @pd.root_validator
+    def bloch_on_both_sides(cls, values):
+        """Error if a Bloch boundary is applied on only one side."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+        num_bloch = isinstance(plus, BlochBoundary) + isinstance(minus, BlochBoundary)
+        if num_bloch == 1:
+            raise SetupError(
+                "Bloch boundaries must be applied either on both sides or on neither side."
+            )
+        return values
+
+    @pd.root_validator
+    def periodic_with_pml(cls, values):
+        """Error if PBC is specified with a PML."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+        num_pbc = isinstance(plus, Periodic) + isinstance(minus, Periodic)
+        num_pml = isinstance(plus, (PML, StablePML, Absorber)) + isinstance(
+            minus, (PML, StablePML, Absorber)
+        )
+        if num_pbc == 1 and num_pml == 1:
+            raise SetupError("Cannot have both PML and PBC along the same dimension.")
+        return values
+
+    @pd.root_validator
+    def periodic_with_pec_pmc(cls, values):
+        """If a PBC is specified along with PEC or PMC on the other side, manually set the PBC
+        to PEC or PMC so that no special treatment of halos is required."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+
+        if isinstance(minus, (PECBoundary, PMCBoundary)) and isinstance(plus, Periodic):
+            plus = minus
+        elif isinstance(plus, (PECBoundary, PMCBoundary)) and isinstance(minus, Periodic):
+            minus = plus
+        values.update({"plus": plus, "minus": minus})
+        return values
+
+    @classmethod
+    def periodic(cls):
+        """Periodic boundary specification on both sides along a dimension.
+
+        Example
+        -------
+        >>> pbc = Boundary.periodic()
+        """
+        plus = Periodic()
+        minus = Periodic()
+        return cls(plus=plus, minus=minus)
+
+    @classmethod
+    def bloch(cls, bloch_vec: Complex):
```

Comment:
complex should be lower case here and in docstring.  we can let user supply `complex` and then it will be converted to `types.Complex` when it is validated in `BlochBoundary`